lsコマンドを作成(他の課題が混ざってしまった為再作成しました)

naoki00413 opened the pull request:
指摘点のファイル間の空白をファイルの最大文字数＋３スペース分に設定しました
並び順を最大3列に対応しました

naoki00413 committed:
指摘点の修正(ファイル間の空白, 並び順を横から縦にして最大3列で表示)

指摘点のファイル間の空白をファイルの最大文字数＋３スペース分に設定しました
並び順を最大3列に対応しました

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,18 @@
+
+def file_data_list
+  file_date = Dir.glob('*') # ファイルデータを取得
+  file_sort = if (file_date.count % 3).zero?
+                file_date.each_slice(file_date.count / 3).to_a # 取得したファイルを３で割る
+              else
+                file_date.each_slice((file_date.count / 3) + 1).to_a # 割り切れない場合
+              end
```
`divmod` という商と余りを同時にとれるメソッドがあるので、それを使うともうちょっとスッキリかけるかもしれません。

https://docs.ruby-lang.org/ja/latest/method/Integer/i/divmod.html

また、条件分岐はできるだけ局所的にしたほうがわかりやすいです。今回の例でいうと、

```ruby
chunk_size = (余りが0のとき) ? (商) : (商 + 1)
chunked_files = filenames.each_slice(chunk_size).to_a
```

みたいな感じです。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+def file_data_list
+  file_date = Dir.glob('*') # ファイルデータを取得
```
変数名をもうちょっとわかりやすくできるとよいですね。
`file_date` だとファイルの日付？と思ってしまうので、何のことだかわかりません。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,18 @@
+
+def file_data_list
+  file_date = Dir.glob('*') # ファイルデータを取得
+  file_sort = if (file_date.count % 3).zero?
```
こちらも `sort` は動詞なので、名前としては `sorted_files` とかのほうがいいと思います。
とはいえ、そもそもこれは `sort` してはないので、 `chunked_files` とか `grouped_files` とか、 `files_for_column` とかになりますかね。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,18 @@
+              else
+                file_date.each_slice((file_date.count / 3) + 1).to_a # 割り切れない場合
+              end
+  vertical_sort = file_sort[0].zip(*file_sort[1..]) # ファイルの並び替え
```
以降も命名をちょっと考えてみてください。
今回のコードだと、実質あまり問題ないですし、書き捨てのコードとかだと適当でもいいんですが、実務で扱うアプリケーションやOSSのライブラリなどのコードになると、書く時間より読まれる時間の方が長いことが多いです。少しでも読みやすくしておくことが生産性の向上につながりますので、今のうちに癖をつけておくとよいかと思います。

yoshitsugu reviewed (commented):
コメントしました！

naoki00413 committed:
指摘点に対応しました(変数名見直しとdivmodメソッドを使用)

divmodメソッドを使用してコードを少しスッキリしました
変数名を見直しました

naoki00413 commented on 04.ls/ls.rb:
```diff
@@ -1,18 +1,16 @@
-  file_space = file_date.max_by(&:length) # ファイルの最大文字数を取得
-  vertical_sort.each do |vertical|
-    vertical.compact.each { print _1.ljust(file_space.length + 3) }
+def file_list
```
変数名をちょこちょこ変更してみました
`chunk_size`と `chunked_files`をそのまま流用させて頂きました🙇‍♂️
その他の変数名は各々のコードが持ってる役割をそのまま変数名にしてみました

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,16 @@
+
+def file_list
+  file_names = Dir.glob('*') # ファイルデータを取得
+  divide_file = file_names.count.divmod(3)
```
Rubyは多重代入ができますので、最初から商と余りをわけておくと便利かなと思います。
https://docs.ruby-lang.org/ja/latest/doc/spec=2foperator.html#multiassign

```ruby
quotient, remainder = file_names.count.divmod(3)
```

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,16 @@
+  divide_file = file_names.count.divmod(3)
+  chunk_size = divide_file[1].zero? ? divide_file[0] : divide_file[0] + 1
+  chunked_files = file_names.each_slice(chunk_size).to_a
+  file_sort = chunked_files[0].zip(*chunked_files[1..]) # ファイルの並び替え
```
これsortはしていないのですよね。sortというと `[1,4,3,2]` という配列を `[1,2,3,4]` にするなど、無秩序なものを秩序だてた順番にする、というイメージがあります。これはただ順番をいれかえているだけなので、 `sort` ではないかなと思います。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,16 @@
+  chunk_size = divide_file[1].zero? ? divide_file[0] : divide_file[0] + 1
+  chunked_files = file_names.each_slice(chunk_size).to_a
+  file_sort = chunked_files[0].zip(*chunked_files[1..]) # ファイルの並び替え
+  space_output = file_names.max_by(&:length) # ファイルの最大文字数を取得
```
これ最大文字数のファイル名をとってきてますよね。コメントも違いそうですし、変数名も違いそうです。

yoshitsugu reviewed (commented):
コメントしました！

naoki00413 committed:
指摘点を修正しました(多重代入の使用,変数名の見直し)

多重代入の使用によりコードをよりスッキリしました
変数名の見直しをしました
file_sort → file_replace, space_output → get_max_length_file

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -0,0 +1,16 @@
+  quotient, remainder = file_names.count.divmod(3)
+  chunk_size = remainder.zero? ? quotient : quotient + 1
+  chunked_files = file_names.each_slice(chunk_size).to_a
+  file_replace = chunked_files[0].zip(*chunked_files[1..]) # ファイルの並び替え
```
動詞でおわるのがやっぱりちょっと違和感があります。
また、Arrayの場合は複数形の名詞にするとわかりやすいです。
説明的に書くと、 `files_rearranged_for_display` みたいな感じですかね。
行列の転置と考えて `transposed_files` とかでもよいかと思います。実際 `transpose` というメソッドがあり、同等の処理ができます。

naoki00413 replied:
レビューありがとうございます！今回は`transposed_files`を流用させて頂きます🙇‍♂️

yoshitsugu reviewed (commented):
コメントしました。

naoki00413 committed:
変数名の変更をしました

変数名をfile_replaceからtransposed_filesに変更しました

yoshitsugu reviewed (approved):
OKです！
